hrjobcontract/CRM/Hrjobcontract/BAO/HRJobContract.php

```diff
@@ -524,32 +524,42 @@ public static function getActiveContracts($startDate = null, $endDate = null)
   }
 
   /**
-   * Return the current contract for the contact if exist.
+   * Return the current revision for current contract for the contact if exist.
+   * which is the contract with current revision start date is
+   * on or before the current date and end date
+   * (is more than or equal the current date) or (null/empty).
+   *
+   * also note :
+   * 1) two contracts can't overlap.
+   * 2) two revision can't have the same effective date (not implement yet -
+   *  but when it does ( TODO: remove create_date DESC from the query ) ).
    *
    * @param int $contactID
    * @return array|null
    */
   public static function getCurrentContract($contactID)  {
-    try  {
-      $queryParam = array(1 => array($contactID, 'Integer'));
-      $query = "SELECT hrjc.id as contract_id , hrjd.*
-                FROM civicrm_hrjobcontract hrjc
-                LEFT JOIN civicrm_hrjobcontract_revision hrjr
-                ON hrjr.jobcontract_id = hrjc.id
-                LEFT JOIN civicrm_hrjobcontract_details hrjd
-                ON hrjr.details_revision_id = hrjd.jobcontract_revision_id
-                WHERE hrjc.contact_id = %1
-                AND hrjr.effective_date <= CURDATE()
-                AND ( hrjr.effective_end_date > CURDATE() OR hrjr.effective_end_date IS NULL)
-                AND ( hrjd.period_end_date > CURDATE() OR hrjd.period_end_date IS NULL)
-                AND hrjc.deleted = 0
-                AND hrjr.deleted = 0
-                LIMIT 1";
-      $response = CRM_Core_DAO::executeQuery($query, $queryParam);
-      $result =  $response->fetch() ? $response : null;
-    } catch(CiviCRM_API3_Exception $ex)  {
-      $result =  null;
+    $currentDate = date('Y-m-d');
+    $queryParam = array(1 => array($contactID, 'Integer'));
+    $query = "SELECT hrjc.id as contract_id , hrjd.*
+      FROM civicrm_hrjobcontract hrjc
+      LEFT JOIN civicrm_hrjobcontract_revision hrjr
+      ON hrjr.jobcontract_id = hrjc.id
+      LEFT JOIN civicrm_hrjobcontract_details hrjd
+      ON hrjr.details_revision_id = hrjd.jobcontract_revision_id
+      WHERE hrjc.contact_id = %1
+      AND hrjr.effective_date <= '{$currentDate}'
+      AND hrjc.deleted = 0
+      AND hrjr.deleted = 0
+      ORDER BY hrjr.effective_date DESC , created_date DESC
+      LIMIT 1";
+    $response = CRM_Core_DAO::executeQuery($query, $queryParam);
+    if ($response->fetch())  {
+      if (empty($response->period_end_date) || $response->period_end_date >= $currentDate )  {
```

Comment:
that won't work , to achieve the same behavior with SQL only I have to write a nested select statement ... Imagine the following scenario to understand why what you are suggesting won't work : 
1- suppose the current day date is 2016-08-13
2 - add a contract with start date = 2016-08-01 and end date = 2016-08-31 . (the effective date will automatically be 2016-08-01)
3- create a new revision with start date = 2016-08-01 and end date = 2016-08-11 (less than today date which 2016-08-13) and set the effective date to (2016-08-04).

now since the new revision effective date = 2016-08-04 so it is the most current revision but since the end date for the contract is 2016-08-11 which is in the past ( remmber that we supposed that today date is 2016-08-13 ) then this user will not have any active contract since the most recent revision end date is in the past .

if I want to update the query according to you suggestion to look like this :

```php
SELECT hrjc.id as contract_id , hrjd.*
      FROM civicrm_hrjobcontract hrjc
      LEFT JOIN civicrm_hrjobcontract_revision hrjr
      ON hrjr.jobcontract_id = hrjc.id
      LEFT JOIN civicrm_hrjobcontract_details hrjd
      ON hrjr.details_revision_id = hrjd.jobcontract_revision_id
      WHERE hrjc.contact_id = %1
      AND hrjr.effective_date <= '{$current_date}'
      AND (hrjd.period_end_date  >= '{$current_date}'OR hrjd.period_end_date  IS NULL)
      AND hrjc.deleted = 0
      AND hrjr.deleted = 0
      ORDER BY hrjr.effective_date DESC , created_date DESC
      LIMIT 1 
 ```

then the output will be the first revision since it meet the criteria ( effective date <= current date .. and end date >= current date ) ... which is not we want , we want it to return null since this contract is not an active contract . to solve this problem I can rewrite it and use nested SQL statements but the current method is just easier and have better performance .